site/src/contexts/ProxyContext.tsx

```diff
@@ -66,41 +97,39 @@ export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
   } = useQuery({
     queryKey,
     queryFn: getWorkspaceProxies,
-    // This onSuccess ensures the local storage is synchronized with the
-    // proxies returned by coderd. If the selected proxy is not in the list,
-    // then the user selection is removed.
-    onSuccess: (resp) => {
-      setAndSaveProxy(proxy.selectedProxy, resp.regions)
-    },
   })
 
-  // Everytime we get a new proxiesResponse, update the latency check
+  // Every time we get a new proxiesResponse, update the latency check
   // to each workspace proxy.
   const proxyLatencies = useProxyLatency(proxiesResp)
 
-  const setAndSaveProxy = (
-    selectedProxy?: Region,
-    // By default the proxies come from the api call above.
-    // Allow the caller to override this if they have a more up
-    // to date list of proxies.
-    proxies: Region[] = proxiesResp?.regions || [],
-  ) => {
-    if (!proxies) {
-      throw new Error(
-        "proxies are not yet loaded, so selecting a proxy makes no sense. How did you get here?",
-      )
-    }
-    const preferred = getPreferredProxy(proxies, selectedProxy)
-    // Save to local storage to persist the user's preference across reloads
-    // and other tabs.
-    savePreferredProxy(preferred)
-    // Set the state for the current context.
-    setProxy(preferred)
-  }
+  // updateProxy is a helper function that when called will
+  // update the proxy being used.
+  const updateProxy = useCallback(() => {
+    // Update the saved user proxy for the caller.
+    setUserSavedProxy(loadUserSelectedProxy())
+    setProxy(
+      getPreferredProxy(
+        proxiesResp?.regions ?? [],
+        // For some reason if I use 'userSavedProxy' here,
+        // the tests fail. It does not load "undefined" after a
+        // clear, but takes the previous value.
+        loadUserSelectedProxy(),
+        proxyLatencies,
+      ),
+    )
+  }, [userSavedProxy, proxiesResp, proxyLatencies])
```

Comment:
Having `userSavedProxy` in the dependency array is going to cause an infinite state loop. We are saying 'rerun the above code every time `userSavedProxy` changes, but at the same time, we are changing its value on [ln 110](https://github.com/coder/coder/pull/7515/files#diff-1beec2c2724792e5a29bc182e07983cd672b52ea138eca517ddb65c5e30bb500R110).
```suggestion
  }, [proxiesResp, proxyLatencies])
```